### ISSUE TITLE
ACL and user management

### ISSUE THREAD

**Issue by Zyles** [dingo/api#600]
Is ACL and user management out of scope for this project?

It is a big part of an API to secure it and sure there are a lot of packages and ways to implement it. But I am kinda wanting to find a blueprint for securing an API just like this project is for building APIs.

A complete "platform" so to speak.

Maybe you can recommend some good packages that would synergize well with Dingo.

**Comment by jasonlewis** [comment]
This package implements authentication providers for Laravel's basic authentication as well as OAuth 2.0 and JWT.

So, that said, all this package really does is provide a mechanism to authenticate requests. What happens thereafter is on you, the developer.

As an aside, the OAuth 2.0 spec does cover scopes. While not exactly the same as ACL they can be used in a similar manner to restrict access to certain endpoints depending on authorization given.

For example, your client-side application may request an OAuth 2.0 token with the `delete_photos` scope. A token is issued with the scope. If a user of your application then attempts to delete a photo you can use this package to check that the token has been authorized with that scope. This might look like the following in your controller:

```php
class PhotoController
{
    use Helpers;

    public function __construct()
    {
        $this->scopes('delete_photos', ['only' => ['destroy']]);
    }
    
    public function destroy($id)
    {
        // only tokens authorized with the "delete_photos" scope will be able to hit this endpoint
    }
}
```

TL;DR: Authentication is included in this package, however, ACL beyond the implemented OAuth 2.0 related stuff is out of scope.
